Basic example of a java sse plugin

LinneaStovring-Nielsen opened the pull request:
### Status
```
[ ] Under development
[x] Waiting for code review
[x] Waiting for merge
```
### Information
```
[ ] Contains breaking changes
[ ] Contains new API(s)
[x] Contains documentation
[ ] Contains test
[x] Contains examples
```
### To-do list
```
[x] [The secure connection in the example is statically linked to OpenSSL and APR-lib. It is possible to do this dynamically, see https://github.com/grpc/grpc-java/blob/master/SECURITY.md. Have anyone done this? ]
```

LinneaStovring-Nielsen committed:
Add first draft of unfinnished java plugin.

Add an sse java plugin to later be used as a basic example.

LinneaStovring-Nielsen committed:
Add (unfinnished) evaluate script

Add evaluate script to the plugin so that it can be used as a basic example
showing how to do script evaluation in a java plugin.

LinneaStovring-Nielsen committed:
Add new functions and the pom.xml file

Add more functions that can be executed to make the example more useful.
Also add the pom.xml file to show which dependencies are needed.

LinneaStovring-Nielsen committed:
Add Cache/NoCache functions

Add functions for showing how the cache can be turned of to be consistent
with other basic examples.

LinneaStovring-Nielsen committed:
Add logging options and secure connection

Add a properties file for the logging so that the plugin user can
set how much logging will be done. Also add a secure conection so
that a secure connection with Qlik Sense Server is possible.

LinneaStovring-Nielsen committed:
Add more documentation and clean upp of code

Add documentaion so that it becomes easier for others
to create their own sse java plugin. Some cleanup to
make the code easier to read.

qlikmats commented on examples/java/PluginServer.java:
```diff
@@ -0,0 +1,547 @@
+                            .parseFrom(metadata.get(Metadata.Key.of("qlik-functionrequestheader-bin", BINARY_BYTE_MARSHALLER))); //Java 8
+                            logger.finest("Function request header.");
+                            logHeader(responseHeaders); //Java 8
+                            if(header.getFunctionId()==5) {
```
spaces surrounding e.g. assignment operator, comparison (<, >, == etc.). I'll leave this one comment, but there are several places throughout the file where this should be fixed.

qlikmats commented on examples/java/PluginServer.java:
```diff
@@ -0,0 +1,547 @@
+        private ServerSideExtension.BundledRows evalScript(ServerSideExtension.ScriptRequestHeader header) {
+            
+            logger.fine("evalScript called");
+            ServerSideExtension.BundledRows.Builder builder = ServerSideExtension.BundledRows.newBuilder();
```
A more descriptive name for builder would be nice, that indicates what is being built when you f.e. see builder.build()

qlikmats commented on examples/java/PluginServer.java:
```diff
@@ -0,0 +1,547 @@
+            return builder.build();
+        }
+        
+        private ServerSideExtension.BundledRows evalScript(ServerSideExtension.ScriptRequestHeader header, ServerSideExtension.BundledRows bundledRows) {
```
Seems like much code in common with previous method. Could it be refactored to avoid code duplication?

qlikmats commented on examples/java/PluginServer.java:
```diff
@@ -0,0 +1,547 @@
+            }
+            
+            //final ServerSideExtension.BundledRows.Builder builder = ServerSideExtension.BundledRows.newBuilder();
+            //&& header.getFunctionType()==ServerSideExtension.FunctionType.SCALAR
```
Remove code that is commented out.

qlikmats reviewed (commented):
In general, without having read the code in detail, I suggest trying to split the code over several classes/files, f.e. server (lifecycle) related code in one file, plugin impl. in another etc. if there is a way to do this that makes sense :)

qlikmats commented on examples/java/README.md:
```diff
@@ -0,0 +1,80 @@
+The plugin is partly asynchronous since several method calls can be made to the plugin at the same time. But a `StreamObserver` (which sends data to and from the plugin) can only handle one call at a time, and they need 
+to be in a specific order, therefore the plugin cannot be fully asynchronous.
+
+## Prerequisites
```
Does this have to be done by the user? Can't we have this structure from the beginning, to make it easier?

qlikmats commented on examples/java/README.md:
```diff
@@ -0,0 +1,80 @@
+**In a command prompt:** go to the grpc-java directory, run `gradlew.bat build` and `gradlew.bat install`. This installs grpc in your local maven repository (usually located at `C:/Users/<user>/.m2`).
+
+## Changing the default logging
+Move the `javapluginlogger.properties` file to the `my-app/target` directory (it needs to be in the same directory as the `my-app-1.0-SNAPSHOT.jar` file).
```
I don't know maven, but I suspect that this file could be moved automatically during the build?

qlikmats reviewed (commented):
I am not sure I remember enough Java, but isn't it best practice to use package declaration(s) and matching folder structure(s)? So, f.e. create a structure (under /src/main/java) like: com/qlik/sse and declare a package com.qlik.sse in PluginServer.java?

qlikmats commented on examples/java/README.md:
```diff
@@ -0,0 +1,80 @@
+In this file you can edit the settings for the logging. There are different [levels](https://docs.oracle.com/javase/7/docs/api/java/util/logging/Level.html) of logging depending on how much 
+information you want (`OFF`, `SEVERE`, `WARNING`, `INFO`, `CONFIG`, `FINE`, `FINER`, `FINEST`, `ALL`).
+
+* `.level`: Sets the default level for all loggers, in this case they are set to info. 
```
I think a short comment about logging, with a reference to the logger documentation is sufficient.

qlikmats commented on examples/java/README.md:
```diff
@@ -0,0 +1,80 @@
+Install Qlik Sense Desktop June release or later.
+To be able to connect to Sense you need to specify which port Sense should look for a plugin at. Go to `Documents/Qlik/Sense` and create or edit the `Settings.ini` file so that it includes 
+the line `SSEPlugin=javaPlugin,localhost:50071`. `javaPlugin` is the name of the plugin. This can be whatever you want but since the name of the plugin in the example app is set to `javaPlugin` the 
+example will not work if you set it to something else (unless you change the name of the plugin in every call to the plugin in the app). `50071` is the port. You should change this to `<Port number>` 
```
Keep it short, like: "'javaPlugin' is an arbitrary name, that needs to match the name used in Sense, when referencing the plugin."

qlikmats commented on examples/java/README.md:
```diff
@@ -0,0 +1,80 @@
+in the `generated-sources` folder.
+
+### If you are having trouble...
+It is possible to install the protobuff and the grpc packages manually if they for some reason are not downloaded by maven when you run `mvn package`. If you need to do this, install the protocol buffer first
```
protobuf

qlikmats commented on examples/java/README.md:
```diff
@@ -0,0 +1,80 @@
+with input parameters if it is not an aggregation and if the script can be evaluated for one row of input parameters independently of the other rows of parameters. The scripts being evaluated however are JavaScript scripts 
+since there is no "eval" function in java.
+
+Synchronous/asynchronous:
```
For me, this section doesn't provide much useful info for the purpose of running an example plugin.

LinneaStovring-Nielsen committed:
Split the code into three separate files

Split the code so that it becomes easier to read and to make
it easier to update it in the future. Update package names
and edit files to follow Java standards.

LinneaStovring-Nielsen commented:
I have now split the code into three files and updated the documentation to reflect that. I have also (hopefully) fixed the problems pointed out by @qlikmats except for the part about storing packages as com/qlik/sse since maven automatically stores them in the local repository when you run mvn package.

kjartan-qlik commented on examples/java/basic_example/src/main/proto/ServerSideExtension.proto:
```diff
@@ -0,0 +1,131 @@
+/*
```
The proto file is already included in the repository as proto/ServerSideExtension.proto. Is it not possible for you to refer to that one instead of including another copy of it?

LinneaStovring-Nielsen replied:
Both yes and no. As I have understood it, standard Maven practice is to include the .proto file like this (and if you add tests you add another copy of it). I can probably find another way to do it. Then it is a question about if it should be possible to download just the basic_example folder or if you need to download everything to run the example?

kjartan-qlik commented on examples/java/basic_example/README.md:
```diff
@@ -0,0 +1,62 @@
+# Running the example
+
+## Prerequisites
+This example was created using maven, so the easiest way to run the example is to install [maven](http://maven.apache.org/ ). 
```
Maven is written with capital 'M'.

kjartan-qlik commented on examples/java/basic_example/README.md:
```diff
@@ -0,0 +1,62 @@
+## Prerequisites
+This example was created using maven, so the easiest way to run the example is to install [maven](http://maven.apache.org/ ). 
+
+There are other ways to do this. If you don't use maven, what you need is the protocol buffer package and the grpc package (in that order) and the protoc 
```
What does 'this' refer to? Building the example?

kjartan-qlik commented on examples/java/basic_example/README.md:
```diff
@@ -0,0 +1,62 @@
+## Generate the java plugin (and the classes from the .proto file)
+**In a command prompt:** go to the `basic_example` directory and run `mvn package`.
+
+Now, got to the `target` directory in the `basic_example` directory. In the `target` folder, you can find a jar file, that is the plugin. The code generated from the `.proto` file can be found 
```
'go to the..'

qlikmats commented on examples/java/basic_example/src/main/java/Plugin/PluginServer.java:
```diff
@@ -0,0 +1,100 @@
+                } catch (Exception e) {
+                    logger.log(Level.WARNING, "Invalid port, using default value: " + port);
+                }
+            } else if (args[i].equals("--pemDir")) {
```
What if --pemDir is given as e.g. "./some/dir"? Since it's concatenated with the .pem file name, the dir needs to have a trailing slash, right? See if there is some path utility library that can handle this, so dir can be given both with and without trailing slash.

LinneaStovring-Nielsen replied:
I hadn't thought about that, fixed now.

qlikmats commented on examples/java/basic_example/src/main/java/Plugin/JavaPlugin.java:
```diff
@@ -0,0 +1,394 @@
+             .setPluginVersion("v1.0.0")
+             .addFunctions(ServerSideExtension.FunctionDefinition.newBuilder()
+                 .setName("HelloWorld")
+                 .setFunctionId(0)
```
Define constants for function IDs since they are also switched upon elsewhere in the code.

qlikmats commented on examples/java/basic_example/src/main/java/Plugin/JavaPlugin.java:
```diff
@@ -0,0 +1,394 @@
+                 switch(functionId) {
+                     case 2: responseObserver.onNext(bundledRowsBuilder.addRows(ServerSideExtension.Row.newBuilder().addDuals(ServerSideExtension.Dual.newBuilder().setNumData(sum(columnSum)))).build());
+                         break;
+                     case 3: responseObserver.onNext(bundledRowsBuilder.addRows(ServerSideExtension.Row.newBuilder().addDuals(ServerSideExtension.Dual.newBuilder().setStrData(stringBuilder.toString()))).build());
```
I think the lines are too long.

qlikmats commented on examples/java/basic_example/src/main/java/Plugin/JavaPlugin.java:
```diff
@@ -0,0 +1,394 @@
+         };
+     }
+     
+     private ServerSideExtension.BundledRows evalScript(ServerSideExtension.ScriptRequestHeader header, ServerSideExtension.BundledRows bundledRows) {
```
Too long line.

qlikmats commented on examples/java/basic_example/src/main/java/Plugin/JavaPlugin.java:
```diff
@@ -0,0 +1,394 @@
+         return bundledRowsBuilder.build();
+     }
+     
+     private boolean evalScript(String script, ServerSideExtension.BundledRows.Builder bundledRowsBuilder, ScriptEngine engine, ServerSideExtension.DataType returnType) {
```
Too long line.

qlikmats commented on examples/java/basic_example/src/main/java/Plugin/PluginServerInterceptor.java:
```diff
@@ -0,0 +1,69 @@
+                    ServerSideExtension.FunctionRequestHeader header = ServerSideExtension.FunctionRequestHeader
+                    .parseFrom(metadata.get(Metadata.Key.of("qlik-functionrequestheader-bin", BINARY_BYTE_MARSHALLER))); //Java 8
+                    logger.finest("Function request header.");
+                    logHeader(responseHeaders); //Java 8
```
What does the "Java 8" comments mean?

LinneaStovring-Nielsen replied:
That you need java 8 for this to work. Should I remove it or make the comment more clear?

qlikmats replied:
Prerequisites already states Java 8 as a requirement, so I think it can be removed.

tobiaslindulf reviewed (commented):
add a .gitignore file in basic_example folder with the following content:

dependency-reduced-pom.xml
/target/*
!/target/javapluginlogger.properties

cjpersson commented:
I was asked to review this (I guess) because we mainly work with Java and that we have made a prototype with a Java SSE. So here are my comments.

It is great to get an example to start with! It will save a lot of time building SSEs using Java.

One thing that is a bit awkward in Java when building SSEs is the way incoming metadata is handled compared to how grpc is integrated in other languages. It would be helpful to explain the purpose of the interceptor and the thread local storage used for metadata. That is a good insight that it takes some time to get when starting building Java SSEs. A comment in the code explaining this mechanism in PluginServerInterceptor and JavaPlugin would be good. 

Folders and package names must match. Normally the structure used here should not compile. I guess the build script does something unconventional to make it compile.

Do not import classes in same package. They are automatically visible.

Use package names starting with a lower case letter. The correct thing to use here would be something like com.qlik.sse.basicExample but for an example it might be easier with folder hierarchies that are not that deep so qlik.basicExample is ok. That would imply that the source files would be in src/main/java/qlik/basicExample.

LinneaStovring-Nielsen commented:
@cjpersson Thank you for your comments! On the metadata explanation, would it be better to have a more detailed explanation in the [documentation](https://github.com/LinneaStovring-Nielsen/server-side-extension/blob/master/examples/java/basic_example/WritingAnSSEPluginUsingJava.md) or do you want that explanation in the code? or both?

qlikmats commented on examples/java/basic_example/src/main/java/com/qlik/sse/basicExample/PluginServer.java:
```diff
@@ -0,0 +1,99 @@
+package com.qlik.sse.basicExample;
```
package names should be all lowercase (https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html)

qlikmats commented on examples/java/basic_example/src/main/java/com/qlik/sse/basicExample/JavaPlugin.java:
```diff
@@ -0,0 +1,405 @@
+package com.qlik.sse.basicExample;
```
package names should be all lowercase (https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html)

qlikmats commented on examples/java/basic_example/src/main/java/com/qlik/sse/basicExample/PluginServerInterceptor.java:
```diff
@@ -0,0 +1,69 @@
+package com.qlik.sse.basicExample;
```
package names should be all lowercase (https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html)

qlikmats reviewed (approved):
Correct the package names when you get an opportunity.
